src/HDF5.jl

```diff
@@ -1712,20 +1712,120 @@ function create_external_dataset(parent::Union{File,Group}, name::AbstractString
-    h5do_write_chunk(dataset, H5P_DEFAULT, UInt32(filter_mask), offs, length(chunk_bytes), chunk_bytes)
+    h5d_write_chunk(dataset, H5P_DEFAULT, UInt32(filter_mask), offs, length(chunk_bytes), chunk_bytes)
+end
+
+"""
+    do_write_chunk(dataset::Dataset, offset, chunk_bytes::Vector{UInt8}, filter_mask=0)
+
+    Write a raw chunk at a given linear index
+    index is 1-based and consecutive up to the number of chunks
+"""
+function do_write_chunk(dataset::Dataset, index::Integer, chunk_bytes::Vector{UInt8}, filter_mask=0)
+    checkvalid(dataset)
+    index -= 1
+    h5d_write_chunk(dataset, H5P_DEFAULT, UInt32(filter_mask), index, chunk_bytes)
+end
+
+"""
+    do_read_chunk(dataset::Dataset, offset)
+
+    Read a raw chunk at a given offset.
+    offset is 1-based of rank `ndims(dataset)` and must fall on a chunk boundary.
+"""
+function do_read_chunk(dataset::Dataset, offset)
+    checkvalid(dataset)
+    offs = collect(hsize_t, reverse(offset)) .- 1
+    filters = Ref{UInt32}()
+    buf = h5d_read_chunk(dataset, offs; filters = filters)
+    return (filters[], buf)
+end
+
+"""
+    do_read_chunk(dataset::Dataset, index::Integer)
+
+    Read a raw chunk at a given index.
+    index is 1-based and consecutive up to the number of chunks
+"""
+function do_read_chunk(dataset::Dataset, index::Integer)
+    checkvalid(dataset)
+    index -= 1
+    filters = Ref{UInt32}()
+    buf = h5d_read_chunk(dataset, index; filters = filters)
+    return (filters[], buf)
 end
 
-struct ChunkStorage
+struct ChunkStorage{I <: IndexStyle, N} <: AbstractArray{Tuple{UInt32,Vector{UInt8}},N}
     dataset::Dataset
 end
+ChunkStorage{I,N}(dataset) where {I,N}= ChunkStorage{I,N}(dataset)
+Base.IndexStyle(::ChunkStorage{I}) where {I <:IndexStyle} = I()
 
-function Base.setindex!(chunk_storage::ChunkStorage, v::Tuple{<:Integer,Vector{UInt8}}, index::Integer...)
+# ChunkStorage{IndexCartesian,N} (default)
+
+function ChunkStorage(dataset)
+    ndims = get_extent_ndims(dataset)
+    ChunkStorage{IndexCartesian, ndims}(dataset)
+end
+
+Base.size(cs::ChunkStorage{IndexCartesian}) = Int64.(get_num_chunks_per_dim(cs.dataset))
+ 
+
+function Base.axes(cs::ChunkStorage{IndexCartesian})
+    chunk = get_chunk(cs.dataset)
+    extent = get_extent_dims(cs.dataset)[1]
+    ntuple(i->1:chunk[i]:extent[i], length(extent))
+end
+
+function Base.setindex!(chunk_storage::ChunkStorage{IndexCartesian}, v::Tuple{<:Integer,Vector{UInt8}}, index::Integer...)
     do_write_chunk(chunk_storage.dataset, hsize_t.(index), v[2], UInt32(v[1]))
 end
 
+function Base.getindex(chunk_storage::ChunkStorage{IndexCartesian}, index::Integer...)
+    do_read_chunk(chunk_storage.dataset, hsize_t.(index))
+end
+
+# ChunkStorage{IndexLinear,1}
+
+ChunkStorage{IndexLinear}(dataset) = ChunkStorage{IndexLinear,1}(dataset)
+Base.size(cs::ChunkStorage{IndexLinear}) = ( Int64( get_num_chunks(cs.dataset) ) ,)
+Base.length(cs::ChunkStorage{IndexLinear}) = Int64( get_num_chunks(cs.dataset) )
```

Comment:
I think
```suggestion
Base.size(cs::ChunkStorage{IndexLinear})   = (Int64(get_num_chunks(cs.dataset)),)
Base.length(cs::ChunkStorage{IndexLinear}) =  Int64(get_num_chunks(cs.dataset))
```
looks better and is more consistent stylistically if you want to maintain the across-line alignment.